### DISCUSSION TITLE
Implement End of the Round automations (including Luck Ends AE duration)

### DISCUSSION THREAD

**Discussion by Savantford** [Savantford/foundry-weirdwizard#250] [New Feature Requests]
https://gitlab.com/foundryvtt_pathfinder1e/foundryvtt-pathfinder1/-/blob/master/module/documents/combat.mjs#L290-297

**Comment by mhilbrunner** [comment]
# Luck ends effects automation

## Rules

Paging through the book, the following require Luck rolls for "luck ends":

- Talents, and their sub-benefits (i.e. some of Fighter's fighting styles)
- Afflictions, mostly to end them, either at end of round or through the Overcome action
- Spells and their effects
- Other effects, for example the health damage-over-time effect on Acid (pg. 63) that is not a normal afflication, but has luck ends; or the "You disappear from existence" effect in the Wild Magic Effects table (pg. 88)

In all these cases, luck ends can be used by the source (e.g. to recharge an ability) or a target (to end an effect).
For reference, this is the full text on "luck ends" (pg. 27):
> [...] you never make a luck roll against a creature or object, the target is always 10.
> (LUCK ENDS): If you are subject to an effect that has the (luck ends) notation, **you make a luck roll at the end of the round. If you have multiple (luck ends) effects affecting you, you make a separate luck roll for each effect, in any order you choose. If you have a (luck ends both) or (luck ends all) effect, you make a single roll.** On a success, the effect ends. On a failure, the effect persists until the end of the next round, at which point you repeat the luck roll to end the effect. You can end a (luck ends) effect with a successful use of the Overcome action.

Especially take note of the bolded part.
Then there's also these general "end of round" rules (pg. 41):
> Some effects begin or end at the end of a turn. If multiple effects begin or end at the end of your turn, you resolve any harmful effects first, in the order you choose, and then other effects, again in the order you choose. [...]
> First, the Sage handles all combatants not controlled by the players. Each combatant resolves all harmful effects first, in any order, and then any remaining effects, again in any order. Then, the players do the same for their characters, after which the combat continues with the start of another round.

## Design thoughts

There broadly seem to be two kinds of things to handle here: effects on a target or self, and *recharging*/*regaining uses*. The latter probably should be handled separately, because there are also other kinds of recharges ("you lose access for... 1 hour/minute", "until after you rest", "you have X charges and regain all charges on Y", etc). So, focusing on effects here first.

As an example, I'm going to use a talent that has the following effect on the target: **The target takes 1D6 damage at the end of each round, and becomes *blinded* and *confused* (luck ends all).**

- The design probably needs to make use of Active Effects.
- Active effects may need a checkbox for "Harmful"
  Alternatively, we could try to derive this automatically: does the effect have any negative key? But not sure that will cover all cases, and not be too finnicky. I guess we could make it a dropdown of "Auto/Harmful/Beneficial" so you could still manually override, not sure.
- Active effects need an effect for "add/subtract health/damage at end of round"
- Active effects need an effect for "add affliction X"
- Active effects already have a "Luck ends" duration that can be set

I'm imagining the following flow:
1. Someone rolls the above talent. That talent has an Active Effect: "1D6 damage end of round, add affliction: blinded, add afliction: confused, duration: luck ends".
2. Somehow, that effect gets put on the target actor. Either by targetting/automation, and/or alternatively (and something that worked well in other systems for me) manually: rolling the talent to chat also adds a link to the embedded AE that can be dragged on target tokens/actors. See https://foundryvtt.com/packages/item-effects-to-chat-5e for an example.
3. During **combat**, at the **end of the round**, we check *luck ends* effects in this order:
3.1 Harmful effects on combatant actors without a player owner
3.2. All other effects on combatant actors without a player owner
3.3. Harmful effects on combatant actors with a player owner
3.4. All other effects on combatant actors with a player owner

As for what to do with each, I can see 3 options:
1. **Roll automatically.** For each luck roll needed, luck is rolled.
2. **Prompt to roll.** For each luck roll needed, the roll dialog is shown (with info which effect is being rolled for).
3. **Inform the player.** A chat card is output, listing all effects that can be rolled for, with a button to open a roll dialog for each.

In each case, on a successful roll, the effect is removed with information output to chat.

I can see this being a per-actor sheet setting (with defaults set as a system or user setting), as I can see all of them being desireable for different actors. Players may initially prefer automatic rolling, some may prefer more control, or have talents/spells/reactions they may want to consider, so they may want to be prompted. Some may want full control and no annoying dialogs popping up each round :-)

The sage may often prefer full automation for NPCs. Likewise, a summoner/necromancer may prefer full automation for summons.

I'm not 100% sure what the intended behaviour for "luck ends" is outside of combat/initiative. Rolls each 3-6 seconds? Automatic success? Automatic success after some time?

**Comment by mhilbrunner** [comment]
I think we're mainly missing:
1. Adding a AE effect "Add affliction" that makes the AE add the given affliction
2. For items that have uses/charges, currently they have a "Recover on rest?" checkbox. That should probably become a multiselect of the following choices (all that I could find in *Shadows*):
    - None (no automatic recovery)
    - Luck ends
    - Recover on start of your next turn
    - Recover on end of your next turn
    - Recover after 1 minute/end of combat
    - Recover after 1 hour
    - Recover after 24 hours
    - Recover after rest
    - Recover after your quest
    - Recover after your next quest
3. Probably may also make sense to have a "Lose access to this while under the following afflictions:" ... configuration on items, then we can later grey out/disable something like the Commander's *Inspiring Commands* when confused/controlled/stunned/unconscious
